### PR TITLE
8268362: [REDO] C2 crash when compile negative Arrays.copyOf length after loop

### PR DESCRIPTION
**Pull request by huishi-hs**
This redos PR https://github.com/openjdk/jdk/pull/4238

Some closed tests fail due to original PR, the main reasons are:

8268325/8268345
CastIINode created in AllocateArrayNode::make_ideal_length apply on a negative array length. GVN transform converts these CastIINode to top. This breaks assumptions, eg., array index node type is TypeInt.
8268345
Some CastIINode is created in LoadRange::Ideal, which doesn't apply GVN transform. CastIINode can not be found in GVN hashtable and cause assertion failure.
Based on previous PR #4238, this PR unifies CastIINode processing post AllocateArrayNode::make_ideal_length. Setup type and add into igvn worklist. Avoid apply GVN transform to break different assumptions in parser.

Based on closed test failrue message, two new tests are added

TestNegArrayLengthAsIndex1 : fail in debug build even without previous PR #4238, it loads a negative array length and apply GVN transform (In GraphKit::load_array_length) which covert length to top node. It also assert in Parse::array_addressing.

TestNegArrayLengthAsIndex2 : similar with failrue in https://bugs.openjdk.java.net/browse/JDK-8268301

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268362](https://bugs.openjdk.java.net/browse/JDK-8268362): [REDO] C2 crash when compile negative Arrays.copyOf length after loop

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Roland Westrelin](https://openjdk.java.net/census#roland) (@rwestrel - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/49/head:pull/49` \
`$ git checkout pull/49`

Update a local copy of the PR: \
`$ git checkout pull/49` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/49/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 49`

View PR using the GUI difftool: \
`$ git pr show -t 49`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/49.diff">https://git.openjdk.java.net/jdk17/pull/49.diff</a>

</details>

### CODE REVIEW

**Comment by dean-long** [comment]
I tested your changes and there were no related failures.

**Comment by huishi-hs** [comment]
> I tested your changes and there were no related failures.

@dean-long  Thanks!

File: src/hotspot/share/opto/library_call.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/library_call.cpp:4460]: Add comment why multiply `CastII` could be here (because each load_array_length() call in guard checks will generate now separate CastII nodes). All these CastII nodes are replaced here with one: original `alloc_length`.

I am wondering if will affect result of guard checks in runtime.

File: src/hotspot/share/opto/library_call.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/library_call.cpp:4513]: Use wrapper method to avoid duplication.

**Review by vnkozlov — COMMENTED** [review]
Few comments.

File: src/hotspot/share/opto/graphKit.cpp
Comment (huishi-hs) [reply src/hotspot/share/opto/graphKit.cpp:1192]: @vnkozlov  Thanks for your review!

GraphKit::load_array_length code is different with new_array() and arraycopy_move_allocation_here(), it doesn't replace_in_map. Currently GraphKit::cast_replace_array_length_post_allocation warpper is used in new_array() and arraycopy_move_allocation_here().

Would you prefer a wrapper for all these three places? It would be like following code.
```
Node* GraphKit::cast_array_length_post_allocation(AllocateArrayNode* alloc, const TypeOopPtr* oop_type, bool replace_map /false in load_array_length/) {
  Node* length = alloc->in(AllocateNode::ALength);
  if (replace_map == false || map()->find_edge(length) >= 0) {
    Node* ccast = alloc->make_ideal_length(oop_type, &_gvn);
    if (ccast != length) {
      _gvn.set_type_bottom(ccast);
      record_for_igvn(ccast);
      if (replace_map) {
        replace_in_map(length, ccast);
      }
      return cast;
    }
  }
  return length;
}
```

File: src/hotspot/share/opto/library_call.cpp
Comment (huishi-hs) [reply src/hotspot/share/opto/library_call.cpp:4463]: done.

Would you share more detail about your concern for "if will affect result of guard checks in runtime."?

File: src/hotspot/share/opto/graphKit.cpp
Comment (vnkozlov) [reply src/hotspot/share/opto/graphKit.cpp:1192]: All places, please. I prefer different names for this method and its parameters `GraphKit::array_ideal_length(alloc , ary_type, replace_length_in_map)`. No need for parameter's comment in declaration - it is clear from name.

**Review by vnkozlov — COMMENTED** [review]
I forgot to say that tests should be in compiler/arraycopy/ directory.

File: src/hotspot/share/opto/graphKit.cpp
Comment (huishi-hs) [reply src/hotspot/share/opto/graphKit.cpp]: done， new wrapper is added, all 3 places: 1) new_array 2) arraycopy_move_allocation_here 3)  load_array_length  use GraphKit::array_ideal_length(alloc , ary_type, replace_length_in_map)

File: src/hotspot/share/opto/library_call.cpp
Comment (huishi-hs) [reply src/hotspot/share/opto/library_call.cpp]: done

**Review by vnkozlov — APPROVED** [review]
Looks good. Let me test it.

**Comment by vnkozlov** [comment]
Testing looks good.

**Comment by huishi-hs** [comment]
> Testing looks good.

@vnkozlov thanks!

**Comment by huishi-hs** [comment]
May I have a second review for this PR?

**Review by rwestrel — APPROVED** [review]
Looks good to me.

**Comment by huishi-hs** [comment]
> Looks good to me.

Thanks again! @rwestrel

**Pull request closed without merging**
